src/Graph.ts

```diff
@@ -135,6 +135,18 @@ export default class Graph {
 		return ast;
 	}
 
+	ensureModule(module: Module | ExternalModule) {
+		// Make sure that the same module / external module can only be added to the
+		// graph once since it may be requested multiple times over the life of a
+		// service.
+		const modules: Array<Module | ExternalModule> =
+			module instanceof Module ? this.modules : this.externalModules;
+
+		if (!modules.includes(module)) {
+			modules.push(module);
+		}
+	}
```

Comment:
Just convert `this.modules` to a `Set` instead of doing this, should also have better performance for large bundles. We are mostly iterating over them, which works just as well for a `Set`, and it is a private variable anyway.